sdk/common/src/main/java/io/opentelemetry/sdk/resources/EntityDetector.java

```diff
@@ -0,0 +1,22 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.sdk.resources;
+
+import java.util.List;
+
+/**
+ * The Entity detector in the SDK is responsible for detecting possible entities that could identify
+ * the SDK (called "associated entities"). For Example, if the SDK is running in a kubernetes pod,
+ * it may provide an Entity for that pod.
+ */
+public interface EntityDetector {
+  /**
+   * Discovers {@link Entity} and their current attributes.
+   *
+   * @return a list of discovered entities.
+   */
+  List<Entity> detectEntities();
```

Comment:
Should a single entity detector return multiple entities or a single entity?

Reply:
Multiple, as is the case today for most non-otel detectors.

E.g. `k8s` entity detector may detect *all* relevant entities (k8s.cluster, k8s.namespace, k8s.pod, etc.).  AWS/GCP/Azure/Other Cloud detector might find `cloud` entity in addition to `host`, `faas`, etc.